packages/fxa-content-server/app/scripts/lib/email-domain-validator.js

```diff
@@ -0,0 +1,116 @@
+ * as display an error/tooltip for the email address field.
+ *
+ * The validation is performed after the existing account email check during
+ * the form submission process. At this point, the email address has gone
+ * through: 1) regex validation, 2) Mailcheck typo suggestion (which the user
+ * can ignore), and 3) existing email address check.
+ *
+ * There are three possible validation results: 'MX', 'A', and 'none'.
+ *  - 'MX': MX record exists, proceed
+ *  - 'A': no MX, but A record exists, display tooltip but allow submission
+ *  - 'none: neither MX or A records found, block submission
+ *
+ * If the validation request itself fails, we allow the submission to
+ * continue.
+ */
+
+import AuthErrors from './auth-errors';
+import Tooltip from '../views/tooltip';
+import TopEmailDomains from './top-email-domains';
+import xhr from './xhr';
+const t = msg => msg;
+
+const EMAIL_VALIDATION_ENDPOINT = '/validate-email-domain';
+// This allows us to distinguish between our own rejection and a server error in
+// a `catch` that'll catch both.
+const REJECTION_CODE = '__INVALID_EMAIL_DOMAIN__';
+const MISTYPED_EMAIL_MESSAGE = t('Mistyped email?');
+
+const resolveDomain = domain =>
+  xhr.ajax({
+    contentType: 'application/json',
+    dataType: 'json',
+    data: { domain },
+    type: 'POST',
+    url: EMAIL_VALIDATION_ENDPOINT,
+  });
+
+const checkEmailDomain = ($element, view) => {
+  const showInvalidDomainError = domain => {
+    const invalidDomainError = AuthErrors.toInvalidEmailDomainError(domain);
+    view.showValidationError($element, invalidDomainError);
+    view.logEvent('emailDomainValidation.fail');
+  };
+
+  const showTooltip = () => {
+    const tooltip = new Tooltip({
+      dismissible: true,
+      extraClassNames: 'tooltip-suggest tooltip-error',
+      invalidEl: $element,
+      type: 'emailDomainValidation',
+      unsafeMessage: MISTYPED_EMAIL_MESSAGE,
+    });
+
+    tooltip.on('destroyed', () =>
+      view.logEvent('emailDomainValidation.dimissed')
+    );
+    tooltip.render();
+    view.logEvent('emailDomainValidation.fallback');
+  };
+
+  return new Promise((resolve, reject) => {
+    const emailAddress = $element.val();
+    const [, domain] = emailAddress.split('@');
+
+    if (TopEmailDomains.has(domain)) {
+      return resolve();
+    }
+
+    // User could repeatedly submit the form
+    if ($element.data('previousDomain') === domain) {
+      // if the previous results is 'A' allow it
+      if ($element.data('previousDomainResult') === 'A') {
+        return resolve();
+      }
+      return reject();
+    }
+
+    view.logEvent('emailDomainValidation.triggered');
+
+    resolveDomain(domain)
+      .then(resp => {
+        const { result } = resp;
+        $element.data('previousDomain', domain);
+        $element.data('previousDomainResult', result);
+
+        if (result === 'MX') {
+          view.logEvent('emailDomainValidation.success');
```

Comment:
Do any of these events need to be reported to Amplitude? If so, you'll need to work with @irrationalagent on the event names/properties, and add some translation [to this list of events in server/lib/amplitude.js](https://github.com/chenba/fxa/blob/6e5714881980035daa086bdd2dddd9674f67654b/packages/fxa-content-server/server/lib/amplitude.js#L40).

Reply:
@davismtl Do we need to report email domain validation results to Amplitude?

Reply:
@shane-tomlinson asks a good question.

I think we could (should?) track failed and successful domain validation on an email first form submit event property. We should look into this. CC @irrationalagent 

I would however not block on this because we can infer a lot from our volume of bounced emails. I would expect that with this enhancement that we will see bounced emails drop.